### PR TITLE
Implemented ERC20Gateway::deposit

### LINKED ISSUE #272: Implement ERC20Gateway::deposit
Implement `deposit` method in `ERC20Gateway`
- Transfers the ERC20 token (value token) from the depositor (msg.sender) address to the `ERC20Gateway` contract.
```
     ERC20Gateway::deposit
        - value token address
        - amount
        - beneficiary address
        - gas price
        - gas limit
```
Emit event: DepositIntentDeclared
```    
DepositIntentDeclared
    - token address
    - amount
    - nonce
    - beneficiary
    - fee gas price
    - fee gas limit
    - depositor
```

### PR DESCRIPTION
**Pull request by kreloaded**
Fixes #272 
- Implemented ERC20Gateway::deposit()
- Implemented unit tests for ERC20Gateway::deposit()
- Implemented event ERC20Gateway::DepositIntentDeclared

### CODE REVIEW

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 1]
```diff
@@ -26,6 +27,21 @@ import "../proxies/MasterCopyNonUpgradable.sol";
  */
 contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
 
+    /* Events */
+
+    /* Emitted when deposit intent hash is declared. */
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
    /** Emitted when deposit intent is declared. */
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 2]
```diff
@@ -110,4 +126,66 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+    /**
+     * @notice Deposit funds to mint on auxiliary chain. Depositor needs to
+     *         approve this contract with the deposit amount.
+     *
+     * @param _valueToken Address of value token.
+     * @param _amount Value token token deposit amount in atto.
+     * @param _beneficiary Address of beneficiary on auxiliary chain.
+     * @param _feeGasPrice Fee gas price at which rewards will be calculated.
+     * @param _feeGasLimit Fee gas limit at which rewards will be calculated.
+     */
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol:161]: Please add \pre and \post conditions as per new guidelines.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 3]
```diff
@@ -110,4 +126,66 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
         emit GatewayProven(messageInbox, _blockNumber);
     }
 
+    /**
+     * @notice Deposit funds to mint on auxiliary chain. Depositor needs to
+     *         approve this contract with the deposit amount.
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]: ` Depositor needs to approve this contract with the deposit amount.`
This is a \pre condition

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 4]
```diff
@@ -110,4 +126,66 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     *
+     * @param _valueToken Address of value token.
+     * @param _amount Value token token deposit amount in atto.
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _amount Amount deposited in atto.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 5]
```diff
@@ -110,4 +126,66 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        address _beneficiary,
+        uint256 _feeGasPrice,
+        uint256 _feeGasLimit
+    )
+        external
+        returns (bytes32 messageHash_)
+    {
+        require(
+            _amount != 0,
+            "Deposit amount should be greater than 0"
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
            "Deposit amount is 0"
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 6]
```diff
@@ -110,4 +126,66 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        );
+        require(
+            _beneficiary != address(0),
+            "Beneficiary address must not be 0."
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
            "Beneficiary address is 0."
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 7]
```diff
@@ -110,4 +126,66 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        );
+        require(
+            _amount > _feeGasPrice.mul(_feeGasLimit),
+            "Deposit amount should be greater than max reward."
+        );
+
+        bytes32 depositIntentHash = hashDepositIntent(_valueToken, _amount, _beneficiary);
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]: This line is going above guideline limit of 79 characters.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 8]
```diff
@@ -110,4 +126,66 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+        address _valueToken,
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _feeGasPrice,
+        uint256 _feeGasLimit
+    )
+        external
+        returns (bytes32 messageHash_)
+    {
+        require(
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol:172]: Add validation that _valueToken address should not be zero

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 9]
```diff
@@ -26,6 +27,21 @@ import "../proxies/MasterCopyNonUpgradable.sol";
...
+    /* Emitted when deposit intent hash is declared. */
+    event DepositIntentDeclared(
+        bytes32 messageHash,
+        address valueToken,
+        uint256 amount,
+        uint256 nonce,
+        address beneficiary,
+        uint256 feeGasPrice,
+        uint256 feeGasLimit,
+        address depositor
```
Comment (abhayks1) [inline contracts/erc20-gateway/ERC20Gateway.sol]: `depositor` can come after `valueToken` in ordering.

File: test/erc20_gateway/erc20_gateway/deposit.js [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+  beforeEach(async () => {
+    erc20Gateway = await ERC20Gateway.new();
+    erc20Cogateway = accountProvider.get();
+    stateRootProvider = accountProvider.get();
+
+    depositor = accountProvider.get();
+    valueToken = await Utils.deployMockToken(depositor, 200);
+
+    console.log('ValueToken address :-', valueToken.address);
```
Comment (abhayks1) [inline test/erc20_gateway/erc20_gateway/deposit.js]: Remove this console.log and multiple new lines.

File: test/erc20_gateway/erc20_gateway/deposit.js [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+      + `found ${depositorBalanceAfterTransfer.toString(10)}`,
+      );
+
+      assert.strictEqual(
+        actualMessageHash,
+        calculatedMessageHash,
+        'Actual message hash must match with calculated message hash.'
+      + `Expected message hash is ${calculatedMessageHash}`
+      + `found ${actualMessageHash}`,
+      );
```
Comment (abhayks1) [inline test/erc20_gateway/erc20_gateway/deposit.js:139]: Also add test for `outbox[messageHash_] = true;` which is set in the contract.

**Review by abhayks1 — DISMISSED** [review]
Nice work 💯 
There are inline comments.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 1 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 2 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol:161]: Added \pre and \post conditions

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 4 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 5 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 6 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 7 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 8 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol:172]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 9 above)
Comment (kreloaded) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done.

File: test/erc20_gateway/erc20_gateway/deposit.js (hunk 10 above)
Comment (kreloaded) [reply test/erc20_gateway/erc20_gateway/deposit.js]: Removed.

File: test/erc20_gateway/erc20_gateway/deposit.js (hunk 11 above)
Comment (kreloaded) [reply test/erc20_gateway/erc20_gateway/deposit.js:139]: Done.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 12]
```diff
@@ -26,6 +27,21 @@ import "../proxies/MasterCopyNonUpgradable.sol";
  */
 contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
 
+    /* Events */
+
+    /** Emitted when deposit intent is declared. */
+    event DepositIntentDeclared(
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol:33]: Please change this to the following:

```
event DepositIntentDeclared(
        uint256 amount,
        uint256 nonce,
        address beneficiary,
        uint256 feeGasPrice,
        uint256 feeGasLimit,
        address depositor,
        address valueToken,
        bytes32 messageHash
    );
```

I will also update the ticket.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 13]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
         emit GatewayProven(messageInbox, _blockNumber);
     }
 
+    /**
+     * @notice Deposit funds to mint on auxiliary chain.
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @notice Deposit ERC20 token to mint utility token on the auxiliary chain.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 14]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     *
+     * @param _valueToken Address of value token.
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _valueToken Address of ERC20 token.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 15]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * @param _amount Amount deposited in atto
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _amount Amount of token to be deposited in atto
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 16]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * @param _beneficiary Address of beneficiary on auxiliary chain.
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _beneficiary Address of beneficiary on the auxiliary chain.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 17]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * @param _feeGasPrice Fee gas price at which rewards will be calculated.
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _feeGasPrice Gas price at which fee will be calculated.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 18]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * @param _feeGasLimit Fee gas limit at which rewards will be calculated.
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```suggestion
     * @param _feeGasLimit Gas limit at which fee will be capped.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 19]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     *
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol:139]: Please add `@returns`

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 2 above)
Comment (deepesh-kn) [reply contracts/erc20-gateway/ERC20Gateway.sol:161]: Pre and post conditions can be as follows:
```
     * \pre  `msg.sender` should approve this contract for `_amount` number of
     *       token transfer.
     * \pre  `_valueToken` address must not be zero.
     * \pre  `_amount` should be greater than max reward. Max reward is
     *       calculated as `_feeGasPrice.mul(_feeGasLimit),`
     * \pre  `_beneficiary` address must not be zero.
     *
     * \post Adds a new entry in `outbox` mapping storage variable. The value is
     *       set as `true` for `messagehash_` in `outbox` mapping. The
     *       `messageHash_` is obtained by calling
     *       `MessageOutbox.declareMessage` with the parameters `depositIntentHash`,
     *      `nonces[msg.sender]`,`_feeGasPrice`,`_feeGasLimit`,`msg.sender`.
     *       depositIntentHash is calculated by calling
     *      `ERC20GatewayBase.hashDepositIntent()` with `_valueToken`,
     *       `_amount` and `_beneficiary` as input parameters.
     * \post Updates the `nonces` storage mapping variable by incrementing the
     *       value for `msg.sender` by one.
     * \post Transfers `_amount` of tokens from `msg.sender` to ERC20Gateway contract.
     * \post Emits `DepositIntentDeclared` event with the address of `messageHash_`,
     *       `_valueToken`, `msg.sender`, `_amount`, `nonce`, `_beneficiary`,
     *       `_feeGasPrice` and `_feeGasLimit` parameters.
```

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 20]
```diff
@@ -110,4 +126,88 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     *       `_feeGasPrice` and `_feeGasLimit` parameters.
+     * \post Transfer approved value tokens from depositor to ERC20Gateway.
+     */
+    function deposit(
+        address _valueToken,
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _feeGasPrice,
+        uint256 _feeGasLimit
+    )
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol:168]: Requesting to update the signature as follows, will update the ticket also.

```
function deposit(
        uint256 _amount,
        address _beneficiary,
        uint256 _feeGasPrice,
        uint256 _feeGasLimit,
        address _valueToken
    )
```

**Review by deepesh-kn — CHANGES_REQUESTED** [review]

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**